app/src/main/java/com/webkeyz/batchtwotraining/FeedDataSource.java

```diff
@@ -0,0 +1,69 @@
+package com.webkeyz.batchtwotraining;
+
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.paging.PageKeyedDataSource;
+
+import com.webkeyz.batchtwotraining.models.ArticlesItem;
+import com.webkeyz.batchtwotraining.models.Feed;
+import com.webkeyz.batchtwotraining.network.FeedApi;
+import com.webkeyz.batchtwotraining.network.ServiceGenerator;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class FeedDataSource extends PageKeyedDataSource<Integer, ArticlesItem> {
+    private static final String TAG = "FeedDataSource";
+    private FeedApi feedApi;
+
+    public FeedDataSource() {
+        feedApi = ServiceGenerator.getFeedApi();
+    }
+
+    @Override
+    public void loadInitial(@NonNull LoadInitialParams<Integer> params, @NonNull final LoadInitialCallback<Integer, ArticlesItem> callback) {
+        feedApi.fetchFeed(Constants.QUERY, Constants.API_KEY, 1, params.requestedLoadSize)
+                .enqueue(new Callback<Feed>() {
+                    @Override
+                    public void onResponse(Call<Feed> call, Response<Feed> response) {
+                        callback.onResult(response.body().getArticles(), null, 2);
+                    }
+
+                    @Override
+                    public void onFailure(Call<Feed> call, Throwable t) {
+                        Log.d(TAG, "onFailure: " + t.getMessage());
```

Comment:
we want to inform view model in case of fail /loading ,  so using network state and invoke this state is important